src/core/Table/hooks/useColumnVisibility.tsx

```diff
@@ -0,0 +1,71 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import { ColumnInstance, HeaderProps, Hooks } from 'react-table';
+import { Checkbox } from '../..';
+import SvgColumnManager from '@itwin/itwinui-icons-react/cjs/icons/ColumnManager';
+import { DropdownMenu } from '../../DropdownMenu';
+import { IconButton } from '../../Buttons/IconButton';
+import { MenuItem } from '../../Menu';
+
+export const COLUMN_MANAGER_ID = 'iui-table-column-manager';
+
+export const useColumnVisibility = <T extends Record<string, unknown>>(
+  hasColumnManager: boolean,
+) => (hooks: Hooks<T>) => {
+  if (!hasColumnManager) {
+    return;
+  }
+
+  hooks.allColumns.push((columns: ColumnInstance<T>[]) => [
+    ...columns,
+    {
+      id: COLUMN_MANAGER_ID,
+      minWidth: 48,
+      width: 48,
+      maxWidth: 48,
+      disableReordering: true,
+      columnClassName: 'iui-slot',
+      cellClassName: 'iui-slot',
+      Header: ({ allColumns }: HeaderProps<T>) => {
+        const [isOpen, setIsOpen] = React.useState(false);
+        const headerCheckBoxes = () =>
+          allColumns
+            .filter(({ id }) => !id.includes('iui-table'))
+            .map((column) => {
+              const { onChange, checked } = column.getToggleHiddenProps();
+              return (
+                <MenuItem
+                  key={column.id}
+                  {...column.getToggleHiddenProps()}
+                  icon={
+                    <Checkbox
```

Comment:
There is also this error in console, which is not nice
![image](https://user-images.githubusercontent.com/81580355/160152070-47d4e0a2-0be4-4bfb-aa51-c7dfa54686c9.png)

Just need to think for the fix.. Maybe you could add `onChange` to checkbox too and then `stopPropagation` in Checkbox `onClick` event, so it does not go up to the `MenuItem`

Reply:
Added an `onChange` event handler to `Checkbox`

Reply:
So few comments here:
- instead of creating mocked event now in `Checkbox` `onChange` handler, you can just pass the original one. (`onChange={onChange}`
- not sure if you have checked, but `stopPropagation` in `onChange` does not help in any way. Because `change` and `click` events are different. So you need to stop it on `onClick` event in `Checkbox`. Very old, but still good debugging way is using `console.log`. Add this in `MenuItem` on click function and see, if clicking checkbox makes it called. What we want to do here is not to make it called, so we dont handle same action twice.

Reply:
This has been updated